src/module_wrap.cc

```diff
@@ -519,50 +568,21 @@ Maybe<URL> ResolveExtensions(const URL& search) {
   return Nothing<URL>();
 }
 
-inline Maybe<URL> ResolveIndex(const URL& search) {
-  return ResolveExtensions<ONLY_VIA_EXTENSIONS>(URL("index", search));
+inline Maybe<URL> ResolveIndex(Environment* env, const URL& search) {
+  return ResolveExtensions<ONLY_VIA_EXTENSIONS>(env, URL("index", search));
 }
 
 Maybe<URL> ResolveMain(Environment* env, const URL& search) {
   URL pkg("package.json", &search);
-  Maybe<uv_file> check = CheckFile(pkg, LEAVE_OPEN_AFTER_CHECK);
-  if (check.IsNothing()) {
-    return Nothing<URL>();
-  }
-
-  Isolate* isolate = env->isolate();
-  Local<Context> context = isolate->GetCurrentContext();
-  std::string pkg_src = ReadFile(check.FromJust());
-  uv_fs_t fs_req;
-  uv_fs_close(nullptr, &fs_req, check.FromJust(), nullptr);
-  uv_fs_req_cleanup(&fs_req);
-
-  // It's not okay for the called of this method to not be able to tell
-  // whether an exception is pending or not.
-  TryCatch try_catch(isolate);
-
-  Local<String> src;
-  if (!String::NewFromUtf8(isolate,
-                           pkg_src.c_str(),
-                           v8::NewStringType::kNormal,
-                           pkg_src.length()).ToLocal(&src)) {
-    return Nothing<URL>();
-  }
 
-  Local<Value> pkg_json;
-  if (!JSON::Parse(context, src).ToLocal(&pkg_json) || !pkg_json->IsObject())
-    return Nothing<URL>();
-  Local<Value> pkg_main;
-  if (!pkg_json.As<Object>()->Get(context, env->main_string())
-                              .ToLocal(&pkg_main) || !pkg_main->IsString()) {
+  PackageConfig pjson = GetPackageConfig(env, pkg.ToFilePath());
+  if (!pjson.exists || !pjson.has_main) {
     return Nothing<URL>();
   }
-  Utf8Value main_utf8(isolate, pkg_main.As<String>());
-  std::string main_std(*main_utf8, main_utf8.length());
-  if (!ShouldBeTreatedAsRelativeOrAbsolutePath(main_std)) {
-    main_std.insert(0, "./");
+  if (!ShouldBeTreatedAsRelativeOrAbsolutePath(pjson.main)) {
+    return Resolve(env, "./" + pjson.main, search);
```

Comment:
Does this get weird with a leading `/`? I'm not entirely sure what this is trying to do versus right below.

Reply:
This is just a rewriting of the original:

```js
if (!ShouldBeTreatedAsRelativeOrAbsolutePath(main_std)) {
  main_std.insert(0, "./");
}
return Resolve(env, main_std, search);
```

into:

```js
if (!ShouldBeTreatedAsRelativeOrAbsolutePath(main_std)) {
  return Resolve(env, "./" + pjson.main, search);
}
return Resolve(env, main_std, search);
```

which I believe is invariant. But happy to revert otherwise too.

Reply:
this looks like a breaking bug that I introduced a long time ago from CJS when I dug up git blame. I'll deal with this in a separate PR.

Reply:
Glad to have helped surface. Ideally I think the `pjson.main` should be pre-sanitized at the cache phase, so that that can already be handled by the time the code gets here.